File: swift/Sources/CantonWalletKit/TransferRegistryClient.swift
import CantonLedgerAPI
import Foundation

/// A registry call failed (non-2xx status or malformed payload).
public struct TransferRegistryError: Error, CustomStringConvertible {
    public let description: String
}

/// Client for a CIP-0056 registry's off-ledger transfer-instruction API
/// (`/registry/transfer-instruction/v1/...`).
///
/// Registries hand out two things a wallet cannot derive on its own: the
/// factory contract to exercise for a new transfer, and per-choice contexts
/// (referenced contracts + disclosed contracts) for accept/reject/withdraw.
public struct TransferRegistryClient: Sendable {
    public let baseURL: URL
    private let session: URLSession

    public init(baseURL: URL, session: URLSession = .shared) {
        self.baseURL = baseURL
        self.session = session
    }

    /// A contract the registry asks us to disclose with the command.
    public struct RegistryDisclosedContract: Sendable {
        public let templateId: String
        public let contractId: String
        public let createdEventBlobBase64: String
        public let synchronizerId: String

        /// As the Ledger API `DisclosedContract` for command submission.
        public func toProto() throws -> Com_Daml_Ledger_Api_V2_DisclosedContract {
            let segments = templateId.split(separator: ":", maxSplits: 2).map(String.init)
            guard segments.count == 3 else {
                throw TransferRegistryError(description: "malformed registry templateId: \(templateId)")
            }
            guard let blob = Data(base64Encoded: createdEventBlobBase64) else {
                throw TransferRegistryError(description: "malformed createdEventBlob for \(contractId)")
            }
            var proto = Com_Daml_Ledger_Api_V2_DisclosedContract()
            proto.templateID.packageID = segments[0]
            proto.templateID.moduleName = segments[1]
            proto.templateID.entityName = segments[2]
            proto.contractID = contractId
            proto.createdEventBlob = blob
            proto.synchronizerID = synchronizerId
            return proto
        }
    }

    public struct RegistryChoiceContext: @unchecked Sendable {
        /// Daml JSON encoding of the ChoiceContext, as parsed JSON (or nil).
        public let choiceContextData: Any?
        public let disclosedContracts: [RegistryDisclosedContract]
    }

    public struct TransferFactory: @unchecked Sendable {
        public let factoryId: String
        /// "self" | "direct" | "offer" — how the registry will route this transfer.
        public let transferKind: String
        public let choiceContext: RegistryChoiceContext
    }

    /// POST `/transfer-factory`: the factory + context for a new transfer.
    public func transferFactory(
        choiceArguments: [String: Any],
        excludeDebugFields: Bool = true
    ) async throws -> TransferFactory {
        let response = try await post(
            path: "registry/transfer-instruction/v1/transfer-factory",
            body: [
                "choiceArguments": choiceArguments,
                "excludeDebugFields": excludeDebugFields,
            ]
        )
        guard
            let factoryId = response["factoryId"] as? String,
            let transferKind = response["transferKind"] as? String
        else {
            throw TransferRegistryError(description: "registry response missing factoryId/transferKind")
        }
        return TransferFactory(
            factoryId: factoryId,
            transferKind: transferKind,
            choiceContext: try choiceContext(response["choiceContext"])
        )
    }

    /// POST `/{id}/choice-contexts/{accept|reject|withdraw}`.
    public func transferInstructionChoiceContext(
        transferInstructionId: String,
        choice: TransferInstructionChoice,
        meta: [String: String] = [:]
    ) async throws -> RegistryChoiceContext {
        var body: [String: Any] = ["excludeDebugFields": true]
        if !meta.isEmpty {
            body["meta"] = meta
        }
        let response = try await post(
            path: "registry/transfer-instruction/v1/\(transferInstructionId)"
                + "/choice-contexts/\(choice.registryPathSegment)",
            body: body
        )
        return try choiceContext(response)
    }

    private func choiceContext(_ json: Any?) throws -> RegistryChoiceContext {
        guard let object = json as? [String: Any] else {
            throw TransferRegistryError(description: "missing choiceContext in registry response")
        }
        let disclosed = try ((object["disclosedContracts"] as? [Any]) ?? []).map { entry in
            guard
                let contract = entry as? [String: Any],
                let templateId = contract["templateId"] as? String,
                let contractId = contract["contractId"] as? String,
                let blob = contract["createdEventBlob"] as? String,
                let synchronizerId = contract["synchronizerId"] as? String
            else {
                throw TransferRegistryError(description: "malformed disclosedContracts entry")
            }
            return RegistryDisclosedContract(
                templateId: templateId,
                contractId: contractId,
                createdEventBlobBase64: blob,
                synchronizerId: synchronizerId
            )
        }
        return RegistryChoiceContext(
            choiceContextData: object["choiceContextData"],
            disclosedContracts: disclosed
        )
    }

    private func post(path: String, body: [String: Any]) async throws -> [String: Any] {
        var request = URLRequest(url: baseURL.appendingPathComponent(path))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await session.data(for: request)
        let status = (response as? HTTPURLResponse)?.statusCode ?? 0
        guard (200..<300).contains(status) else {
            let text = String(data: data, encoding: .utf8) ?? ""
            throw TransferRegistryError(description: "HTTP \(status) from \(path): \(text)")
        }
        guard let object = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw TransferRegistryError(description: "registry response from \(path) is not a JSON object")
        }
        return object
    }
}

File: kotlin/canton-wallet-sdk/src/main/kotlin/io/github/vsima/canton/wallet/TransferRegistryClient.kt
// Copyright (c) 2026 Victor Sima
// SPDX-License-Identifier: Apache-2.0

package io.github.vsima.canton.wallet

import com.daml.ledger.api.v2.CommandsOuterClass
import com.daml.ledger.api.v2.ValueOuterClass
import com.google.protobuf.ByteString
import java.io.IOException
import java.util.Base64
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import kotlinx.serialization.json.putJsonObject
import okhttp3.Call
import okhttp3.Callback
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response

/** A registry call failed (non-2xx status or malformed payload). */
public class TransferRegistryException(message: String) : RuntimeException(message)

/**
 * Client for a CIP-0056 registry's off-ledger transfer-instruction API
 * (`/registry/transfer-instruction/v1/...`).
 *
 * Registries hand out two things a wallet cannot derive on its own: the
 * factory contract to exercise for a new transfer, and per-choice contexts
 * (referenced contracts + disclosed contracts) for accept/reject/withdraw.
 */
public class TransferRegistryClient(
    baseUrl: String,
    private val http: OkHttpClient = OkHttpClient(),
) {
    private val baseUrl = baseUrl.trimEnd('/')

    /** A contract the registry asks us to disclose with the command. */
    public data class RegistryDisclosedContract(
        val templateId: String,
        val contractId: String,
        val createdEventBlobBase64: String,
        val synchronizerId: String,
    ) {
        /** As the Ledger API `DisclosedContract` for command submission. */
        public fun toProto(): CommandsOuterClass.DisclosedContract {
            val segments = templateId.split(":", limit = 3)
            if (segments.size != 3) {
                throw TransferRegistryException("malformed registry templateId: $templateId")
            }
            return CommandsOuterClass.DisclosedContract.newBuilder()
                .setTemplateId(
                    ValueOuterClass.Identifier.newBuilder()
                        .setPackageId(segments[0])
                        .setModuleName(segments[1])
                        .setEntityName(segments[2])
                )
                .setContractId(contractId)
                .setCreatedEventBlob(
                    ByteString.copyFrom(Base64.getDecoder().decode(createdEventBlobBase64))
                )
                .setSynchronizerId(synchronizerId)
                .build()
        }
    }

    public data class RegistryChoiceContext(
        val choiceContextData: JsonElement?,
        val disclosedContracts: List<RegistryDisclosedContract>,
    )

    public data class TransferFactory(
        val factoryId: String,
        /** "self" | "direct" | "offer" — how the registry will route this transfer. */
        val transferKind: String,
        val choiceContext: RegistryChoiceContext,
    )

    /** POST `/transfer-factory`: the factory + context for a new transfer. */
    public suspend fun transferFactory(
        choiceArguments: JsonObject,
        excludeDebugFields: Boolean = true,
    ): TransferFactory {
        val body = buildJsonObject {
            put("choiceArguments", choiceArguments)
            put("excludeDebugFields", excludeDebugFields)
        }
        val response = post("$baseUrl/registry/transfer-instruction/v1/transfer-factory", body)
        return TransferFactory(
            factoryId = response.requireString("factoryId"),
            transferKind = response.requireString("transferKind"),
            choiceContext = choiceContext(response["choiceContext"]),
        )
    }

    /** POST `/{id}/choice-contexts/{accept|reject|withdraw}`. */
    public suspend fun transferInstructionChoiceContext(
        transferInstructionId: String,
        choice: TransferInstructionChoice,
        meta: Map<String, String> = emptyMap(),
    ): RegistryChoiceContext {
        val body = buildJsonObject {
            if (meta.isNotEmpty()) {
                putJsonObject("meta") { meta.forEach { (k, v) -> put(k, v) } }
            }
            put("excludeDebugFields", true)
        }
        val response = post(
            "$baseUrl/registry/transfer-instruction/v1/$transferInstructionId" +
                "/choice-contexts/${choice.registryPathSegment}",
            body,
        )
        return choiceContext(response)
    }

    private fun choiceContext(json: JsonElement?): RegistryChoiceContext {
        val obj = json as? JsonObject
            ?: throw TransferRegistryException("missing choiceContext in registry response")
        val disclosed = (obj["disclosedContracts"] as? JsonArray ?: JsonArray(emptyList())).map {
            val contract = it as? JsonObject
                ?: throw TransferRegistryException("malformed disclosedContracts entry: $it")
            RegistryDisclosedContract(
                templateId = contract.requireString("templateId"),
                contractId = contract.requireString("contractId"),
                createdEventBlobBase64 = contract.requireString("createdEventBlob"),
                synchronizerId = contract.requireString("synchronizerId"),
            )
        }
        return RegistryChoiceContext(
            choiceContextData = obj["choiceContextData"],
            disclosedContracts = disclosed,
        )
    }

    private fun JsonObject.requireString(key: String): String =
        (get(key) as? JsonPrimitive)?.content
            ?: throw TransferRegistryException("registry response missing $key")

    private suspend fun post(url: String, body: JsonObject): JsonObject {
        val request = Request.Builder()
            .url(url)
            .post(body.toString().toRequestBody("application/json".toMediaType()))
            .build()
        val text = http.newCall(request).await(url)
        return Json.parseToJsonElement(text) as? JsonObject
            ?: throw TransferRegistryException("registry response from $url is not a JSON object")
    }

    private suspend fun Call.await(url: String): String =
        suspendCancellableCoroutine { continuation ->
            enqueue(object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    continuation.resumeWithException(e)
                }

                override fun onResponse(call: Call, response: Response) {
                    response.use {
                        val text = it.body?.string().orEmpty()
                        if (!it.isSuccessful) {
                            continuation.resumeWithException(
                                TransferRegistryException("HTTP ${it.code} from $url: $text")
                            )
                        } else {
                            continuation.resume(text)
                        }
                    }
                }
            })
            continuation.invokeOnCancellation { cancel() }
        }
}
